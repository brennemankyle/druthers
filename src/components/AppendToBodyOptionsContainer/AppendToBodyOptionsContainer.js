import React, { forwardRef } from 'react'
import Container from '../Container/Container'
import AppPropTypes from '../../utils/AppPropTypes'

let AppendToBodyOptionsContainer = (props, ref) => {
  let {className, children, ...otherProps} = props

  return <Container className={props.className} ref={ref} {...otherProps}>{props.children}</Container>
}

AppendToBodyOptionsContainer = forwardRef(AppendToBodyOptionsContainer)

AppendToBodyOptionsContainer.propTypes = {
  styles: AppPropTypes.styles.isRequired,
}

export default AppendToBodyOptionsContainer
