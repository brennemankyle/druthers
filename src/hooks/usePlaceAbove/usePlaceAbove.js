// eslint-disable-next-line
import React, { useState, useEffect } from 'react'

let usePlaceAbove = (rect, optionsWrapperRect, skip = []) => {
  const [placeOptionsAbove, setPlaceOptionsAbove] = useState(false)

  // if no space below, place above
  useEffect(() => {
    let windowHeight = window.innerHeight + window.scrollY
    let belowDiff = windowHeight - (rect.y + rect.height + optionsWrapperRect.height)
    let aboveDiff = rect.y - optionsWrapperRect.height
    setPlaceOptionsAbove(belowDiff < 0 && aboveDiff > belowDiff)

    // eslint-disable-next-line
  }, [rect, optionsWrapperRect].concat(skip))

  return placeOptionsAbove
}

export default usePlaceAbove
