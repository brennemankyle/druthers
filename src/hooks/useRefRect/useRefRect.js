// eslint-disable-next-line
import React, { useState, useEffect } from "react";

let getRect = current => {
  if (!current)
    return {
      top: 0,
      bottom: 0,
      left: 0,
      right: 0,
      width: 0,
      height: 0,
      x: -window.innerWidth,
      y: -window.innerHeight
    };

  let { top, bottom, left, right, width, height } = current.getClientRects()[0];

  let x = current.offsetLeft;
  let y = current.offsetTop;

  return { top, bottom, left, right, width, height, x, y };
};

let useRefRect = (ref, skip = []) => {
  const [rect, setRect] = useState(getRect());

  useEffect(() => {
    setRect(getRect(ref.current));
    // eslint-disable-next-line
  }, Object.values(getRect(ref.current)).concat(skip));

  return rect;
};

export default useRefRect;
