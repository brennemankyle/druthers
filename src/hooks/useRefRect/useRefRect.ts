// eslint-disable-next-line
import React, { useState, useEffect } from "react";

export interface DomRect {
  top: number;
  bottom: number;
  left: number;
  right: number;
  height: number;
  y: number;
  width: number;
  x: number;
}

function getRect(current: HTMLElement | undefined | null): DomRect {
  if (!current)
    return {
      top: 0,
      bottom: 0,
      left: 0,
      right: 0,
      width: 0,
      height: 0,
      x: -window.innerWidth,
      y: -window.innerHeight,
    };

  let { top, bottom, left, right, width, height } = current.getClientRects()[0];

  let x = current.offsetLeft;
  let y = current.offsetTop;

  return { top, bottom, left, right, width, height, x, y };
}

function useRefRect(
  ref: React.MutableRefObject<HTMLElement | undefined | null> | null,
  skip: string[] = []
): DomRect {
  const [rect, setRect] = useState(getRect(undefined));

  useEffect(() => {
    setRect(getRect(ref?.current));
    // eslint-disable-next-line
  }, Object.values(getRect(ref?.current)).concat(skip));

  return rect;
}

export default useRefRect;
