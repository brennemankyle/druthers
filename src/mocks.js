import React from 'react'

let noop = () => {}

let mockStyles = {
  colors: {}
}

let mockItem = {
  value: '42',
  label: 'Meaning of life',
}

let mockItemList = [
  {value: '1', label: 'Item 1'},
  {value: '2', label: 'Item 2'},
  {value: '3', label: 'Item 3'},
  {value: '4', label: 'Item 4'},
  {value: '5', label: 'Item 5'},
]

let mockEvent = {
  preventDefault: noop
}

let mockElement = (props) => <div data-props={props} />
let mockInput = (props) => <input data-props={props} />

export {
  mockStyles,
  mockItem,
  mockItemList,
  mockEvent,
  mockElement,
  mockInput, }
