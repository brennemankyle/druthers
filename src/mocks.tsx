import React from "react";
import defaultProps from "./utils/defaultProps";
import { withKeys, noop } from "./utils/utils";

let mockStyles = {
  styles_width: 600,
  styles_multiple: false,
  styles_disabled: false,
  styles_hasSelection: false,
  styles_hasOptions: true,
  styles_optionHighlighted: [],
  styles_rightToLeft: false,
  ...withKeys(defaultProps, "styles_"),
};

let mockItem = {
  value: "42",
  label: "Meaning of life",
};

let mockItemList = [
  { value: "1", label: "Item 1" },
  { value: "2", label: "Item 2" },
  { value: "3", label: "Item 3" },
  { value: "4", label: "Item 4" },
  { value: "5", label: "Item 5" },
];

let mockEvent = {
  preventDefault: noop,
};

let MockElement = (props: any) => <div data-props={props} />;
let MockInput = (props: any) => <input type="text" data-props={props} />;
let MockSvg = () => (
  <svg>
    <path d="M0 0L5 5Z" />
  </svg>
);

let mockRect = {
  top: 10,
  bottom: 110,
  left: 20,
  right: 220,
  width: 200,
  height: 100,
  x: 15,
  y: 25,
};

export {
  mockStyles,
  mockItem,
  mockItemList,
  mockEvent,
  MockElement,
  MockInput,
  MockSvg,
  mockRect,
};
