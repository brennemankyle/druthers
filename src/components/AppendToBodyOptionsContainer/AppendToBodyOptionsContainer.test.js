import React from 'react'
import { shallow, mount } from 'enzyme'
import toJson from 'enzyme-to-json'
import { mockStyles, mockElement, mockItemList } from '../../mocks'
import AppendToBodyOptionsContainer from './AppendToBodyOptionsContainer'
import useRefRect from '../../hooks/useRefRect/useRefRect'
import usePlaceAbove from '../../hooks/usePlaceAbove/usePlaceAbove'

jest.mock('../../hooks/useRefRect/useRefRect', () => {
  return () => {
    return {
      "bottom": 80,
      "height": 80,
      "left": 0,
      "right": 200,
      "top": 0,
      "width": 200,
      "x": 0,
      "y": 600,
    }
  }
})

jest.mock('../../hooks/usePlaceAbove/usePlaceAbove', () => {
  return () => {
    return true
  }
})

global.console.error = jest.fn()

it('renders', () => {
  const parentRef = React.createRef()
  const wrapper = mount(<AppendToBodyOptionsContainer
    styles={mockStyles}
    className="test"
    parentRef={parentRef}
    filteredOptions={mockItemList}
    StyledAppendToBodyOptionsContainer={mockElement}><mockElement /></AppendToBodyOptionsContainer>)

  expect(toJson(wrapper)).toMatchSnapshot()
})

it('forwards ref', () => {
  const parentRef = React.createRef()
  const wrapper = shallow(<AppendToBodyOptionsContainer
    styles={mockStyles}
    className="test"
    parentRef={parentRef}
    filteredOptions={[]}
    StyledAppendToBodyOptionsContainer={mockElement} />)

  expect(parentRef.current).toBeDefined()
})
