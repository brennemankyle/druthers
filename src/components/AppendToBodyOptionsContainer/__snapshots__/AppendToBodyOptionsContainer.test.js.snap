// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<AppendToBodyOptionsContainer
  StyledAppendToBodyOptionsContainer={[Function]}
  className="test"
  filteredOptions={
    Array [
      Object {
        "label": "Item 5",
        "value": "5",
      },
      Object {
        "label": "Item 4",
        "value": "4",
      },
      Object {
        "label": "Item 3",
        "value": "3",
      },
      Object {
        "label": "Item 2",
        "value": "2",
      },
      Object {
        "label": "Item 1",
        "value": "1",
      },
    ]
  }
  parentRef={
    Object {
      "current": null,
    }
  }
  styles={
    Object {
      "colors": Object {},
    }
  }
  updateOn={Array []}
>
  <mockElement
    className="test"
    parentRect={
      Object {
        "bottom": 80,
        "height": 80,
        "left": 0,
        "right": 200,
        "top": 0,
        "width": 200,
        "x": 0,
        "y": 600,
      }
    }
    placeOptionsAbove={true}
    styles={
      Object {
        "colors": Object {},
      }
    }
  >
    <div
      data-props={
        Object {
          "children": <mockElement
            itemList={
              Array [
                Object {
                  "label": "Item 5",
                  "value": "5",
                },
                Object {
                  "label": "Item 4",
                  "value": "4",
                },
                Object {
                  "label": "Item 3",
                  "value": "3",
                },
                Object {
                  "label": "Item 2",
                  "value": "2",
                },
                Object {
                  "label": "Item 1",
                  "value": "1",
                },
              ]
            }
          />,
          "className": "test",
          "parentRect": Object {
            "bottom": 80,
            "height": 80,
            "left": 0,
            "right": 200,
            "top": 0,
            "width": 200,
            "x": 0,
            "y": 600,
          },
          "placeOptionsAbove": true,
          "styles": Object {
            "colors": Object {},
          },
        }
      }
    />
  </mockElement>
</AppendToBodyOptionsContainer>
`;
